File: apps/api/app/recurrences/application/run_due_recurrences.py
from __future__ import annotations

from datetime import date

from app.recurrences.domain.schedule import due_occurrences
from app.recurrences.ports.movement_writer import MovementWriter
from app.recurrences.ports.repository import RecurrenceRepository


class RunDueRecurrences:
    """Materializes every occurrence that has come due into a real movement.

    Idempotent: the repository's last_generated_on cursor means a second run on
    the same day creates nothing.
    """

    def __init__(self, repository: RecurrenceRepository, writer: MovementWriter) -> None:
        self._repository = repository
        self._writer = writer

    def __call__(self, today: date) -> int:
        generated = 0
        for recurrence in self._repository.list_active():
            occurrences = due_occurrences(recurrence, today)
            for occurrence in occurrences:
                self._writer.create(
                    kind=recurrence.kind,
                    amount_cents=recurrence.money.amount_cents,
                    currency=recurrence.money.currency,
                    category_id=recurrence.category_id,
                    occurred_on=occurrence,
                    note=recurrence.note,
                )
                generated += 1
            if occurrences:
                self._repository.mark_generated(recurrence.id, occurrences[-1])
        return generated

File: apps/api/internal/recurrences/application/run_due_recurrences.go
package application

import (
	"context"
	"time"

	"caudal-api/internal/recurrences/domain"
	"caudal-api/internal/recurrences/ports"
)

type RunDueRecurrences struct {
	repository ports.Repository
	writer     ports.MovementWriter
}

func NewRunDueRecurrences(repository ports.Repository, writer ports.MovementWriter) RunDueRecurrences {
	return RunDueRecurrences{repository: repository, writer: writer}
}

func (useCase RunDueRecurrences) Execute(ctx context.Context, today time.Time) (int, error) {
	recurrences, err := useCase.repository.ListActive(ctx)
	if err != nil {
		return 0, err
	}
	generated := 0
	for _, recurrence := range recurrences {
		count, err := useCase.materialize(ctx, recurrence, today)
		if err != nil {
			return generated, err
		}
		generated += count
	}
	return generated, nil
}

func (useCase RunDueRecurrences) materialize(ctx context.Context, recurrence domain.Recurrence, today time.Time) (int, error) {
	occurrences := domain.DueOccurrences(recurrence, today)
	for _, occurrence := range occurrences {
		err := useCase.writer.Create(ctx, recurrence, occurrence)
		if err != nil {
			return 0, err
		}
	}
	if len(occurrences) > 0 {
		if err := useCase.repository.MarkGenerated(ctx, recurrence.ID, occurrences[len(occurrences)-1]); err != nil {
			return 0, err
		}
	}
	return len(occurrences), nil
}
